pyingest/parsers/jats.py

```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/env python
+
+import sys
+import json
+import codecs
+from adsputils import u2asc
+from default import BaseXmlToDictParser
+from xmltodict import unparse
+
+class NoSchemaException(Exception):
+    pass
+
+class WrongSchemaException(Exception):
+    pass
+
+class UnparseableException(Exception):
+    pass
+
+class JATSParser(BaseXmlToDictParser):
+
+    def __init__(self):
+        pass
+
+    def resource_dict(self, fp, **kwargs):
+        d = self.xmltodict(fp, **kwargs)
+# this returns the root of a JATS article, which is the element <article>
+        r = d.get('article',{})
+        return r
+
+    def _attribs(self, r, **kwargs):
+        kl = self._dict(r).keys()
+        attribs = {}
+        for k in kl:
+            if k[0] == '@':
+                attribs[k] = r[k]
+        return attribs
+
+    def _gettext(self, r):
+        if isinstance(r,list):
+            return "; ".join(self._gettext(x) for x in r)
+        elif isinstance(r,tuple):
+            return ": ".join(self._gettext(x) for x in r)
+        elif isinstance(r,dict):
+            return self._gettext(r.items())
+        elif isinstance(r,basestring):
+            return r
+
+
+
+    def parse(self, fp, **kwargs):
+
+        output_metadata=dict()
+
+# The storage of the raw jats data is a kludge to deal with cases where you
```

Comment:
This implementation makes me very uneasy mainly for two reasons:
1. It relies on string to parse xml rather than using a proper parser; one can come up with examples where this will fail under the proper circumstances
2. It takes the XML contents inside the abstract or title without checking for what additional structured content may be there.  This means that we may be accepting additional XML or HTML markup which ADS classic (and SOLR) is not prepared for (suppose there is a <script src="http://evilsite.com/hack.js> in there).

I think we should bite the bullet and use a more sophisticated parser which allows us to examine the contents of these elements, validating and transforming content found in there.
We should probably draft a requirements document before proceeding.

Reply:
I just found there's an xmltodict.unparse, so let me investigate whether that can handle these better than re-reading the raw xml.

That would not fix the issue you raised though, about possible dangerous items in the xml.  It's entirely possible someone could put a malicious item inside an equation or other MML object.

Reply:
I briefly looked into unparse() and it doesn't do the job (does not retain all the correct order and structure of subelements).
FWIW on the classic side, we implemented this kind of checking and filtering in ADS::Abstracts::Checker (see "perldoc ADS::Abstract::Checker").  I imagine we'll want something similar in python to control what gets parsed in.

Reply:
I'll think about this.  If I can find a way to get it to ignore the <inline-formula> namespace or to transmit everything inside it literally without parsing, it could work.  There are certainly perl/cgi modules that can sterilize input, and I'd be surprised if something didn't already exist in python.  It would be unfortunate to lose the MathML, because it is rendered and made sensible to the end user in the browser. 

By the way, the other two fixes are straightforward -- I've already implemented the first, and am working on the second.